### DISCUSSION TITLE
Direct "googlevideo.com" cdn link

### DISCUSSION THREAD

**Discussion by akashdeep000** [fent/node-ytdl-core#1045] [Q&A] [answered]
I can't find a way to get "googlevideo.com"  links.
I have seen many youtube downloader websites just grab the "googlevideo.com" cdn url and show download buttom with those links.
By doing this we  haven't proxy YouTube video and stream through our own server.

**Comment by imaperson1060** [comment, marked as answer]
The download links can be found by doing
```js
function getVideoInfo(id) {
    var videoInfo = await yt.dl.getInfo(id);
    var hd;
    try { hd = videoInfo.formats.find(x => x.itag == 22).url; } catch (e) {}
    var sd = videoInfo.formats.find(x => x.itag == 18).url;
    var audio = videoInfo.formats.find(x => x.itag == 140).url;
    var formats = { hd, sd, audio };
    return formats;
}
```

**Reply by akashdeep000** [reply]
Yeah ☺️ it's work 👍

```js
const ytdl = require('ytdl-core');

async function getVideoInfo(id) {
    var videoInfo = await ytdl.getInfo(id);
    var hd;
    try { hd = videoInfo.formats.find(x => x.itag == 22).url; } catch (e) {}
    var sd = videoInfo.formats.find(x => x.itag == 18).url;
    var audio = videoInfo.formats.find(x => x.itag == 140).url;
    var formats = { hd, sd, audio };
    return formats;
}

getVideoInfo("bR8sE9ubyTI")
.then(results => console.log(results))
```

**Reply by imaperson1060** [reply]
You're welcome 😄

**Comment by AhmedO9473** [comment]
@imaperson1060,
This answer only works for JS that's in a Node JS enviorment, which is server-side, If I'm developing a client side JS, what's the suitable copy of this code:
```js
function getVideoInfo(id) {
    var videoInfo = await yt.dl.getInfo(id);
    var hd;
    try { hd = videoInfo.formats.find(x => x.itag == 22).url; } catch (e) {}
    var sd = videoInfo.formats.find(x => x.itag == 18).url;
    var audio = videoInfo.formats.find(x => x.itag == 140).url;
    var formats = { hd, sd, audio };
    return formats;
}
```

for a client-side JS??

**Reply by imaperson1060** [reply]
It should be the same, provided you can find a ytdl library for the front end. All that code does is it filters through the ytdl output.

If it's the library you're looking for, I don't think there is one, as YouTube's CORS policy makes it impossible. tl;dr, that means that no other domain can make requests to YouTube, requiring the library to be backend so it simulates a "user".

**Comment by AhmedO9473** [comment]
@imaperson1060,
The example doesn't work with a React app, whenever i include (ytdl-core) it says:

> Module not found: Error: Can't resolve 'http' in 'C:\Users\ahma2\apcwytpadfree\node_modules\miniget\dist'
> BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
> This is no longer the case. Verify if you need this module and configure a polyfill for it.
> 
> If you want to include a polyfill, you need to:
>         - add a fallback 'resolve.fallback: { "http": require.resolve("stream-http") }'
>         - install 'stream-http'
> If you don't want to include a polyfill, you can use an empty module like this:
>         resolve.fallback: { "http": false }
> WARNING in ./node_modules/miniget/dist/index.js
> Module Warning (from ./node_modules/source-map-loader/dist/cjs.js):
> Failed to parse source map from 'C:\Users\ahma2\apcwytpadfree\node_modules\miniget\src\index.ts' file: Error: ENOENT: no such file or directory, open 'C:\Users\ahma2\apcwytpadfree\node_modules\miniget\src\index.ts'
> 
> WARNING in ./node_modules/m3u8stream/dist/dash-mpd-parser.js
> Module Warning (from ./node_modules/source-map-loader/dist/cjs.js):
> Failed to parse source map from 'C:\Users\ahma2\apcwytpadfree\node_modules\m3u8stream\src\dash-mpd-parser.ts' file: Error: ENOENT: no such file or directory, open 'C:\Users\ahma2\apcwytpadfree\node_modules\m3u8stream\src\dash-mpd-parser.ts'
> 
> WARNING in ./node_modules/m3u8stream/dist/index.js
> Module Warning (from ./node_modules/source-map-loader/dist/cjs.js):
> Failed to parse source map from 'C:\Users\ahma2\apcwytpadfree\node_modules\m3u8stream\src\index.ts' file: Error: ENOENT: no such file or directory, open 'C:\Users\ahma2\apcwytpadfree\node_modules\m3u8stream\src\index.ts'
> 
> WARNING in ./node_modules/m3u8stream/dist/m3u8-parser.js
> Module Warning (from ./node_modules/source-map-loader/dist/cjs.js):
> Failed to parse source map from 'C:\Users\ahma2\apcwytpadfree\node_modules\m3u8stream\src\m3u8-parser.ts' file: Error: ENOENT: no such file or directory, open 'C:\Users\ahma2\apcwytpadfree\node_modules\m3u8stream\src\m3u8-parser.ts'
> 
> WARNING in ./node_modules/m3u8stream/dist/parse-time.js
> Module Warning (from ./node_modules/source-map-loader/dist/cjs.js):
> Failed to parse source map from 'C:\Users\ahma2\apcwytpadfree\node_modules\m3u8stream\src\parse-time.ts' file: Error: ENOENT: no such file or directory, open 'C:\Users\ahma2\apcwytpadfree\node_modules\m3u8stream\src\parse-time.ts'
> 
> WARNING in ./node_modules/m3u8stream/dist/queue.js
> Module Warning (from ./node_modules/source-map-loader/dist/cjs.js):
> Failed to parse source map from 'C:\Users\ahma2\apcwytpadfree\node_modules\m3u8stream\src\queue.ts' file: Error: ENOENT: no such file or directory, open 'C:\Users\ahma2\apcwytpadfree\node_modules\m3u8stream\src\queue.ts'
> 
> WARNING in ./node_modules/sax/lib/sax.js 139:13-37
> Module not found: Error: Can't resolve 'stream' in 'C:\Users\ahma2\apcwytpadfree\node_modules\sax\lib'
> 
> BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
> This is no longer the case. Verify if you need this module and configure a polyfill for it.
> 
> If you want to include a polyfill, you need to:
>         - add a fallback 'resolve.fallback: { "stream": require.resolve("stream-browserify") }'
>         - install 'stream-browserify'
> If you don't want to include a polyfill, you can use an empty module like this:
>         resolve.fallback: { "stream": false }
> 
> WARNING in [eslint]
> src\App.js
>   Line 1:8:    'logo' is defined but never used     no-unused-vars
>   Line 11:49:  Expected '===' and instead saw '=='  eqeqeq
>   Line 12:47:  Expected '===' and instead saw '=='  eqeqeq
>   Line 13:50:  Expected '===' and instead saw '=='  eqeqeq
> 
> src\Ytapi.js
>   Line 2:7:   'stream' is assigned a value but never used  no-unused-vars
>   Line 7:49:  Expected '===' and instead saw '=='          eqeqeq
>   Line 8:47:  Expected '===' and instead saw '=='          eqeqeq
>   Line 9:50:  Expected '===' and instead saw '=='          eqeqeq
> 
> ERROR in ./node_modules/miniget/dist/index.js 8:31-46
> Module not found: Error: Can't resolve 'http' in 'C:\Users\ahma2\apcwytpadfree\node_modules\miniget\dist'
> 
> BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
> This is no longer the case. Verify if you need this module and configure a polyfill for it.
> 
> If you want to include a polyfill, you need to:
>         - add a fallback 'resolve.fallback: { "http": require.resolve("stream-http") }'
>         - install 'stream-http'
> If you don't want to include a polyfill, you can use an empty module like this:
>         resolve.fallback: { "http": false }
> 
> ERROR in ./node_modules/miniget/dist/index.js 9:32-48
> Module not found: Error: Can't resolve 'https' in 'C:\Users\ahma2\apcwytpadfree\node_modules\miniget\dist'
> 
> BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
> This is no longer the case. Verify if you need this module and configure a polyfill for it.
> 
> If you want to include a polyfill, you need to:
>         - add a fallback 'resolve.fallback: { "https": require.resolve("https-browserify") }'
>         - install 'https-browserify'
> If you don't want to include a polyfill, you can use an empty module like this:
>         resolve.fallback: { "https": false }
> 
> ERROR in ./node_modules/miniget/dist/index.js 10:17-34
> Module not found: Error: Can't resolve 'stream' in 'C:\Users\ahma2\apcwytpadfree\node_modules\miniget\dist'
> 
> BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
> This is no longer the case. Verify if you need this module and configure a polyfill for it.
> 
> If you want to include a polyfill, you need to:
>         - add a fallback 'resolve.fallback: { "stream": require.resolve("stream-browserify") }'
>         - install 'stream-browserify'
> If you don't want to include a polyfill, you can use an empty module like this:
>         resolve.fallback: { "stream": false }
> 
> ERROR in ./node_modules/m3u8stream/dist/dash-mpd-parser.js 11:17-34
> Module not found: Error: Can't resolve 'stream' in 'C:\Users\ahma2\apcwytpadfree\node_modules\m3u8stream\dist'
> 
> BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
> This is no longer the case. Verify if you need this module and configure a polyfill for it.
> 
> If you want to include a polyfill, you need to:
>         - add a fallback 'resolve.fallback: { "stream": require.resolve("stream-browserify") }'
>         - install 'stream-browserify'
> If you don't want to include a polyfill, you can use an empty module like this:
>         resolve.fallback: { "stream": false }
> 
> ERROR in ./node_modules/m3u8stream/dist/index.js 8:17-34
> Module not found: Error: Can't resolve 'stream' in 'C:\Users\ahma2\apcwytpadfree\node_modules\m3u8stream\dist'
> 
> BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
> This is no longer the case. Verify if you need this module and configure a polyfill for it.
> 
> If you want to include a polyfill, you need to:
>         - add a fallback 'resolve.fallback: { "stream": require.resolve("stream-browserify") }'
>         - install 'stream-browserify'
> If you don't want to include a polyfill, you can use an empty module like this:
>         resolve.fallback: { "stream": false }
> 
> ERROR in ./node_modules/m3u8stream/dist/m3u8-parser.js 6:17-34
> Module not found: Error: Can't resolve 'stream' in 'C:\Users\ahma2\apcwytpadfree\node_modules\m3u8stream\dist'
> 
> BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
> This is no longer the case. Verify if you need this module and configure a polyfill for it.
> 
> If you want to include a polyfill, you need to:
>         - add a fallback 'resolve.fallback: { "stream": require.resolve("stream-browserify") }'
>         - install 'stream-browserify'
> If you don't want to include a polyfill, you can use an empty module like this:
>         resolve.fallback: { "stream": false }
> 
> ERROR in ./node_modules/safe-buffer/index.js 3:13-30
> Module not found: Error: Can't resolve 'buffer' in 'C:\Users\ahma2\apcwytpadfree\node_modules\safe-buffer'
> 
> BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
> This is no longer the case. Verify if you need this module and configure a polyfill for it.
> 
> If you want to include a polyfill, you need to:
>         - add a fallback 'resolve.fallback: { "buffer": require.resolve("buffer/") }'
>         - install 'buffer'
> If you don't want to include a polyfill, you can use an empty module like this:
>         resolve.fallback: { "buffer": false }
> 
> ERROR in ./node_modules/ytdl-core/lib/cache.js 3:4-21
> Module not found: Error: Can't resolve 'timers' in 'C:\Users\ahma2\apcwytpadfree\node_modules\ytdl-core\lib'
> 
> BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
> This is no longer the case. Verify if you need this module and configure a polyfill for it.
> 
> If you want to include a polyfill, you need to:
>         - add a fallback 'resolve.fallback: { "timers": require.resolve("timers-browserify") }'
>         - install 'timers-browserify'
> If you don't want to include a polyfill, you can use an empty module like this:
>         resolve.fallback: { "timers": false }
> 
> ERROR in ./node_modules/ytdl-core/lib/index.js 1:20-49
> Module not found: Error: Can't resolve 'stream' in 'C:\Users\ahma2\apcwytpadfree\node_modules\ytdl-core\lib'
> 
> BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
> This is no longer the case. Verify if you need this module and configure a polyfill for it.
> 
> If you want to include a polyfill, you need to:
>         - add a fallback 'resolve.fallback: { "stream": require.resolve("stream-browserify") }'
>         - install 'stream-browserify'
> If you don't want to include a polyfill, you can use an empty module like this:
>         resolve.fallback: { "stream": false }
> 
> ERROR in ./node_modules/ytdl-core/lib/info-extras.js 2:11-33
> Module not found: Error: Can't resolve 'querystring' in 'C:\Users\ahma2\apcwytpadfree\node_modules\ytdl-core\lib'
> 
> BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
> This is no longer the case. Verify if you need this module and configure a polyfill for it.
> 
> If you want to include a polyfill, you need to:
>         - add a fallback 'resolve.fallback: { "querystring": require.resolve("querystring-es3") }'
>         - install 'querystring-es3'
> If you don't want to include a polyfill, you can use an empty module like this:
>         resolve.fallback: { "querystring": false }
> 
> ERROR in ./node_modules/ytdl-core/lib/info.js 1:20-42
> Module not found: Error: Can't resolve 'querystring' in 'C:\Users\ahma2\apcwytpadfree\node_modules\ytdl-core\lib'
> 
> BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
> This is no longer the case. Verify if you need this module and configure a polyfill for it.
> 
> If you want to include a polyfill, you need to:
>         - add a fallback 'resolve.fallback: { "querystring": require.resolve("querystring-es3") }'
>         - install 'querystring-es3'
> If you don't want to include a polyfill, you can use an empty module like this:
>         resolve.fallback: { "querystring": false }
> 
> ERROR in ./node_modules/ytdl-core/lib/info.js 8:4-21
> Module not found: Error: Can't resolve 'timers' in 'C:\Users\ahma2\apcwytpadfree\node_modules\ytdl-core\lib'
> 
> BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
> This is no longer the case. Verify if you need this module and configure a polyfill for it.
> 
> If you want to include a polyfill, you need to:
>         - add a fallback 'resolve.fallback: { "timers": require.resolve("timers-browserify") }'
>         - install 'timers-browserify'
> If you don't want to include a polyfill, you can use an empty module like this:
>         resolve.fallback: { "timers": false }
> 
> ERROR in ./node_modules/ytdl-core/lib/sig.js 1:20-42
> Module not found: Error: Can't resolve 'querystring' in 'C:\Users\ahma2\apcwytpadfree\node_modules\ytdl-core\lib'
> 
> BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
> This is no longer the case. Verify if you need this module and configure a polyfill for it.
> 
> If you want to include a polyfill, you need to:
>         - add a fallback 'resolve.fallback: { "querystring": require.resolve("querystring-es3") }'
>         - install 'querystring-es3'
> If you don't want to include a polyfill, you can use an empty module like this:
>         resolve.fallback: { "querystring": false }
> 
> ERROR in ./node_modules/ytdl-core/lib/sig.js 4:11-24
> Module not found: Error: Can't resolve 'vm' in 'C:\Users\ahma2\apcwytpadfree\node_modules\ytdl-core\lib'
> 
> BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
> This is no longer the case. Verify if you need this module and configure a polyfill for it.
> 
> If you want to include a polyfill, you need to:
>         - add a fallback 'resolve.fallback: { "vm": require.resolve("vm-browserify") }'
>         - install 'vm-browserify'
> If you don't want to include a polyfill, you can use an empty module like this:
>         resolve.fallback: { "vm": false }
> 
> ERROR in [eslint]
> src\App.js
>   Line 9:32:  'id' is not defined  no-undef
> 
> Search for the keywords to learn more about each error.
> 
> webpack compiled with 15 errors and 8 warnings

How do I solve this?

**Reply by imaperson1060** [reply]
This appears to be a problem with the `miniget` module.

**Comment by AhmedO9473** [comment]
I ran the library on a React application and had this error:

> Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://www.youtube.com/get_video_info?video_id=KymaDBqZvfA&c=TVHTML5&cver=7.20210622.10.00&eurl=https%3A%2F%2Fyoutube.googleapis.com%2Fv%2FKymaDBqZvfA&ps=default&gl=US&hl=en&html5=1. (Reason: CORS header ‘Access-Control-Allow-Origin’ missing).

Since every Node.JS app must have a URL, how do I solve this??

**Reply by AhmedO9473** [reply]
That also, it has a 405 (Method Not Allowed) code besides that.

**Reply by imaperson1060** [reply]
Yes, that's how CORS works. Node apps don't inherently have URLs, because you can run them without creating a web server. When you send a request from a Node app's backend, it appears as your IP address, and looks like a normal user is requesting the URL. If you send the request from an HTML page, then the URL will be localhost, or whatever domain you're using, and that is blocked.
